torch/_overrides.py

```diff
@@ -0,0 +1,242 @@
+        methods, in the order in which they should be called.
+
+    """
+    # Runtime is O(num_arguments * num_unique_types)
+    overloaded_types = []
+    overloaded_args = []
+    for arg in relevant_args:
+        arg_type = type(arg)
+        # We only collect arguments if they have a unique type, which ensures
+        # reasonable performance even with a long list of possibly overloaded
+        # arguments.
+        if (arg_type not in overloaded_types and
+                hasattr(arg_type, '__torch_function__')):
+
+            # Create lists explicitly for the first type (usually the only one
+            # done) to avoid setting up the iterator for overloaded_args.
+            if overloaded_types:
+                overloaded_types.append(arg_type)
+                # By default, insert argument at the end, but if it is
+                # subclass of another argument, insert it before that argument.
+                # This ensures "subclasses before superclasses".
+                index = len(overloaded_args)
+                for i, old_arg in enumerate(overloaded_args):
+                    if issubclass(arg_type, type(old_arg)):
+                        index = i
+                        break
+                overloaded_args.insert(index, arg)
+            else:
+                overloaded_types = [arg_type]
+                overloaded_args = [arg]
+
+    return overloaded_types, overloaded_args
+
+
+def implement_torch_function(
+        implementation, public_api, relevant_args, args, kwargs):
+    """Implement a function with checks for __torch_function__ overrides.
+
+    Arguments
+    ---------
+    implementation : function
+        Function that implements the operation on ``torch.Tensor`` without
+        overrides when called like ``implementation(*args, **kwargs)``.
+    public_api : function
+        Function exposed by the public torch API originally called like
+        ``public_api(*args, **kwargs)`` on which arguments are now being
+        checked.
+    relevant_args : iterable
+        Iterable of arguments to check for __torch_function__ methods.
+    args : tuple
+        Arbitrary positional arguments originally passed into ``public_api``.
+    kwargs : tuple
+        Arbitrary keyword arguments originally passed into ``public_api``.
+
+    Returns
+    -------
+    Result from calling `implementation()` or an `__torch_function__`
+    method, as appropriate.
+
+    Raises
+    ------
+    TypeError : if no implementation is found.
+
+    """
+    # Check for __torch_function__ methods.
+    types, overloaded_args = get_overloaded_types_and_args(relevant_args)
+    # Short-cut for common cases: no overload or only Tensor overload
+    # (directly or with subclasses that do not override __torch_function__).
+    if (not overloaded_args or types == _TENSOR_ONLY or
+            all(type(arg).__torch_function__ is _TORCH_FUNCTION
+                for arg in overloaded_args)):
```

Comment:
Hmm. I suppose the benchmarks for the C++ will eventually show up, but the short cut doesn't seem all that short-cutty to me. In particular, `get_overloaded_types_and_args` will treat Tensor as an "overload", so if `__torch_function__` is defined on Tensor (which it is), then you will never actually be in a situation where `overloaded_args` is falsish. (If this code is just meant to be semantics, as opposed to code that will be directly transliterated to C++, you can disregard this comment)

Reply:
You're right, this is not meant for direct translation to C++. There we want to hook in only after the check that inputs are tensor instances. In the `torch.functional` Python functions there's no such check though, so the inputs are checked upfront.